keps/sig-network/2433-topology-aware-hints/README.md

```diff
@@ -0,0 +1,617 @@
+routing at zone level but could be expanded to include region.
+
+In the short term, this is taking the place of two closely related KEPs that
+were never implemented. These KEPs relate to EndpointSlice subsetting and are
+still relevant, just deferred to a later point in time. For more info on this
+transition refer to the following resources:
+
+- [Doc: Updates to Topology in Kubernetes
+  1.21](https://docs.google.com/document/d/1ZzUoFY1SrdjVefl7gVOJZJLt1I1LHttw8pcX95nlgMY/edit)
+- [KEP 2004: Topology Aware
+  Subsetting](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/2004-topology-aware-subsetting).
+- [KEP 2030: Topology Aware
+  Proxying](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/2030-topology-aware-proxying).
+
+## Motivation
+
+Kubernetes clusters are increasingly deployed in multi-zone environments.
+Network traffic is routed randomly to any endpoint matching a Service. Some
+users might want the traffic to stay in the same zone for the following
+reasons:
+- Cost savings: Keeping traffic within a zone can limit cross-zone networking
+  costs.
+- Performance: Traffic within a zone usually has less latency and bandwidth
+  constraints, having a better performance than traffic leaving the zone.
+
+In this KEP we are going to focus on avoiding cross-zone traffic when in-zone
+endpoints would suffice. We're attempting to provide a simple and more automatic
+approach to topology aware routing. This API will still allow users to indicate
+that they prefer to keep traffic in the same zone if there's sufficient
+capacity. With this approach users won't have to configure anything by default
+for most use cases.
+
+### Goals
+- Provide a simple way for users to indicate their preference for keeping
+  traffic in zone.
+- Use the standard topology label `topology.kubernetes.io/zone` to derive the
+  zones of nodes and endpoints.
+- Use EndpointSlice hints as the primary mechanism for topology aware routing.
+- Minimize churn of EndpointSlices while doing topology aware distribution.
+- Minimize the number of new EndpointSlices required.
+- Provide a simple API that requires minimal configuration for most users.
+
+### Non-Goals
+- Real-time distribution rebalancing based on traffic load or distribution
+  feedback or metrics.
+- Multi-cluster topology aware routing (this same pattern may be useful there
+  though).
+- Region based topology aware routing (this may come later).
+- Ensuring that Pods are distributed evenly across zones.
+
+## Proposal
+
+When this feature is enabled, the EndpointSlice controller will be updated to
+provide hints for each endpoint. These hints will initially be limited to a
+single zone per-endpoint. Kube-Proxy will then use these hints to filter the
+endpoints they should route to.
+
+For example, for a Service with 3 endpoints, the EndpointSlice controller may
+create an EndpointSlice with endpoints that look like this:
+
+```
+- addresses: ["10.1.2.3"]
```

Comment:
nit addeses are the same for all endpoints :)

Reply:
Woops, thanks! Should be fixed now.